odl/discr/partition.py

```diff
@@ -0,0 +1,561 @@
+﻿# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Partitons of interval products based on tensor grids.
+
+A partition of a set is a finite collection of nonempty, pairwise
+disjoint subsets whose union is the original set. The partitions
+considered here are based on hypercubes, i.e. the tensor products
+of partitions of intervals.
+"""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+
+from future import standard_library
+standard_library.install_aliases()
+from builtins import object, range, super, zip
+
+# External module imports
+import numpy as np
+
+# ODL imports
+from odl.discr.grid import TensorGrid, RegularGrid
+from odl.set.domain import IntervalProd
+from odl.util.utility import array1d_repr
+
+
+__all__ = ('RectPartition', 'uniform_partition')
+
+_POINT_POSITIONS = ('center', 'left')
+
+
+class RectPartition(object):
+
+    """Rectangular partition by hypercubes based on `TensorGrid`s.
+
+    In 1d, a partition of an interval is implicitly defined by a
+    collection of points x[0], ..., x[N-1] (a grid) which are chosen to
+    lie in the center of the subintervals. The i-th subinterval is thus
+    given by
+    ::
+        I[i] = [(x[i-1]+x[i])/2, (x[i]+x[i+1])/2]
+    """
+
+    def __init__(self, grid, begin=None, end=None, **kwargs):
```

Comment:
I simply thought that this guy should be quite "rough" and simply take an `IntervalProd` and a `TensorGrid`. Any magical creation of things from grids or using only some of the points in begin etc could be moved to a creation utility. In 98% of the use cases users will want something like:

```python
>>> uniform_partition(0, 1, 10) # Gives a partition of the interval [0, 1] into 10 equaly large sub-intervals
```

That guy could then possibly take some advanced parameters (such as the begin_axes logic here).